nnvm/src/pass/plan_memory.cc

```diff
@@ -39,6 +40,22 @@ static int GetDTypeSize(int type_flag) {
   }
 }
 
+static bool BitwiseCompatibleTypes(int type1, int type2) {
+  static std::set<std::pair<int, int>> compatible_pairs { {kUint8, kInt8},
```

Comment:
It seems this set is designed for cast op. For others, it may not suitable. For example, saturate_cast from kUint8 to kInt8 will change trunk data. Relu from kInt8 to kUint8 will also  break this.

Reply:
From my point of view, cast shouldn't have identity attributes because for most of the usecase, cast will modify trunk data. It's abuse of using of identity for working around redundant operations. This should be implemented as a graph pass that detect and delete those redundant operations, including cast with same dtype, elewise_add with 0, relu on uint8, bn with kernel=[1,1] and stride=[1,1], etc. 

Otherwise, we have to add identity to all those operators, it's not maintainable. Also, identity can only ensure memory inplace between input and output. Extra check have to be added in each opeators to avoid the computation, making common usecase even slower. Another problem is, the memory copy between previous GPU operator and next identity CPU operator can't be avoided, making identity useless for this case. Overall,  this is imperfect solution.